packages/core/src/storages/request_queue_v2.ts

```diff
@@ -192,67 +289,87 @@ export class RequestQueue extends RequestProvider {
         }
 
         // We want to fetch ahead of time to minimize dead time
-        if (this.queueHeadIds.length() > 1 && !(await this.hasPendingForefrontRequests())) {
+        if (this.queueHeadIds.length() > 1 && !this.shouldCheckForForefrontRequests) {
             return;
         }
 
-        this._listHeadAndLockPromise ??= this._listHeadAndLock().finally(() => {
-            this._listHeadAndLockPromise = null;
+        this.listHeadAndLockPromise ??= this._listHeadAndLock().finally(() => {
+            this.listHeadAndLockPromise = null;
         });
 
-        await this._listHeadAndLockPromise;
+        await this.listHeadAndLockPromise;
+    }
+
+    private async giveUpLock(id?: string, uniqueKey?: string) {
+        if (id === undefined) {
+            return;
+        }
+
+        try {
+            await this.client.deleteRequestLock(id);
+        } catch {
+            this.log.debug('Failed to delete request lock', { id, uniqueKey });
+        }
     }
 
     private async _listHeadAndLock(): Promise<void> {
-        const forefront = await this.hasPendingForefrontRequests();
+        // Make a copy so that we can clear the flag only if the whole method executes after the flag was set
+        // (i.e, it was not set in the middle of the execution of the method)
+        const shouldCheckForForefrontRequests = this.shouldCheckForForefrontRequests;
 
         const headData = await this.client.listAndLockHead({
-            limit: Math.min(forefront ? this.assumedForefrontCount : 25, 25),
+            limit: 25,
             lockSecs: this.requestLockSecs,
         });
 
+        this.queueHasLockedRequests = headData.queueHasLockedRequests;
+
         const headIdBuffer = [];
+        const forefrontHeadIdBuffer = [];
 
         for (const { id, uniqueKey } of headData.items) {
-            // Queue head index might be behind the main table, so ensure we don't recycle requests
-            if (
-                !id ||
-                !uniqueKey ||
-                this.recentlyHandledRequestsCache.get(id) ||
-                // If we tried to read new forefront requests, but another client appeared in the meantime, we can't be sure we'll only read our requests.
-                // To retain the correct queue ordering, we rollback this head read.
-                (forefront && headData.hadMultipleClients)
-            ) {
-                this.log.debug(`Skipping request from queue head as it's potentially invalid or recently handled`, {
+            if (!id || !uniqueKey) {
+                this.log.warning(`Skipping request from queue head as it's evidently invalid`, {
```

Comment:
I love the subtle jab but let's not give that to users if this path is somehow ever reached.

Hell, is this something that can EVER be reached? I don't think I've seen this happen in what, 3 years?

```suggestion
                this.log.warning(`Skipping request from queue head as it's invalid. Please report this with the provided metadata!`, {
```

Reply:
Yeah, your iteration is the best of the three :smile: I'd love to say that this can't happen, but the check has to be there for a reason. Though I believe that the reason is some memory-storage bug.

Reply:
This check predates me so 🤷‍♂️